unread_ops: Add confirm_dialog to bulk mark unread in interleaved views.

Niloth-p opened the pull request:
Fixes #31292.

[CZO thread](https://chat.zulip.org/#narrow/stream/137-feedback/topic/unintentionally.20marking.20messages.20as.20unread/near/1638015)

Known number: (All Fetched)
- N < 50 - no confirmation dialog
- N >= 50 - confirmation dialog with the number of messages

Unknown number:
- N < 10 - no confirmation dialog
- N >= 10 - confirmation dialog with the minimum number of messages, rounded down to the nearest multiple of 25.

**Screenshots and screen captures:**
[Screen recording for the known number cases](https://github.com/user-attachments/assets/dcdf611d-0d6b-4538-b779-193df443ef4e)
[Screen recording for the unknown number confirmation dialog](https://github.com/user-attachments/assets/eaff8556-93d1-4c9a-a000-7eaf9eda3a60)

Screenshot for the unknown number case (and its threshold)
![image](https://github.com/user-attachments/assets/662f5984-1f1c-4b99-b44a-9f6be87cdb95)


<details>
<summary>Self-review checklist</summary>

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Niloth-p commented on web/src/unread_ops.ts:
```diff
@@ -256,31 +258,71 @@ export function mark_as_unread_from_here(
-                    (!include_anchor && msg.id > message_id),
-            )
-            .map((msg) => msg.id);
+    const rounding_factor = 25;
```
Added these as constants, but found other values like the 200 (in `do_mark_unread()`) hard-coded in the same function, so I'm not sure if this is overdone.

Joelute replied:
Is there a CZO discussion on this? I think it would be helpful to start one to discuss these values with folks on this.
If there's already an CZO discussion please do link it to the PR description.

Niloth-p replied:
The CZO discussion is linked in the first line of the issue that this PR solves.
I've now edited the PR description to include the CZO link now, like you requested.

timabbott replied:
Constants near the top of the file, a la `INITIAL_BATCH_SIZE`, are often good, but generally require more specific names; `threshold_fetched` is probably too generic for that.

Niloth-p reviewed (commented):
Self-reviewed some aspects.

I could not find any existing automated tests for `mark_as_unread_from_here()`. I haven't added or edited any tests for the introduced behavior.

afeefuddin commented on web/src/unread_ops.ts:
```diff
@@ -256,31 +258,71 @@ export function mark_as_unread_from_here(
+    function do_mark_unread(): void {
+        // If we have already fully fetched the current view, we can
+        // send the server the set of IDs to update, rather than
+        // updating on the basis of the narrow.
+        if (is_all_fetched && message_ids_to_update!.length < 200) {
```
Use `assert` from `minimalistic-assert`, or maybe you can do `message_ids_to_update !== undefined` instead of using `is_all_fetched`

afeefuddin reviewed (commented):
I don't know if the logic can be improved, but this looks good to me, apart from a few comments.

Niloth-p commented:
Thank you, @afeefuddin!
Updated to fix the typo (thanks for catching that!) and revert back to using `message_ids_to_update !== undefined` instead of `is_all_fetched`.

Niloth-p commented:
@zormit Could you please review this too?
@zulipbot add "mentor review"

timabbott commented on web/src/unread_ops.ts:
```diff
@@ -256,31 +258,60 @@ export function mark_as_unread_from_here(
-                    (!include_anchor && msg.id > message_id),
-            )
-            .map((msg) => msg.id);
+    const is_all_fetched = message_lists.current.data.fetch_status.has_found_newest();
```
The boolean variable is good, but why not call it `has_found_newest`? `is_all_fetched` can be misleading, since it sounds like the oldest might be fetched too.

timabbott commented on web/src/unread_ops.ts:
```diff
@@ -256,31 +258,71 @@ export function mark_as_unread_from_here(
+    const threshold_fetched = 50;
+    const threshold_unfetched = 10;
+    const is_all_fetched = message_lists.current.data.fetch_status.has_found_newest();
+    const is_in_interleaved_view = !narrow_state.narrowed_by_reply();
```
This check is incorrect; `filter.is_conversation_view() || filter.is_conversation_view_with_near()` is closer to what you want. (near/with views are not accounted for correctly, which have extra operators whose only impact is on the scroll position, not which messages match).

But I guess actually even that check would be wrong: A narrow that has `sender/channel/topic` filters would not be interleaved, because it's a subset of a given conversation.

So maybe you want `current_filter.can_bucket_by("channel", "topic") || filter.can_bucket_by("dm"))` as the check here; that seems like it precisely asks if this contains a single conversation.

Probably reasonable to make that value available as an exported function `filter.is_interleaved_view` in a prep commit, since it took me a while to figure out how to define it.

timabbott commented on web/src/unread_ops.ts:
```diff
@@ -256,31 +258,71 @@ export function mark_as_unread_from_here(
+        if (is_all_fetched) {
+            display_count = message_count.toString();
+        } else if (message_count < threshold_unfetched) {
+            display_count = "fewer-than-threshold";
```
We should consider this information in a separate variable, rather than trying to jam it into the field for what we'll display.

Niloth-p commented:
Update:
- Moved the hardcoded constants to the top of the file, and renamed them. Added a comment explaining their purpose.
- Renamed `is_all_fetched` variable to `has_found_newest`.
- Fixed the `is_interleaved_view` check. Sorry about that. I looked into the narrowing and filtering part of the codebase a bit better now, but I might have still directly checked the `term_types` instead of using `can_bucket_by`, so thank you for the solution.
	- Added that as a preparatory commit, with the `is_interleaved_view()` and corresponding assertions in the test file.
	- Created a const `current_filter` and replaced `message_lists.current.data.filter`'s usage. This doesn't particularly relate to this commit's purpose of introducing a confirmation dialog. Should this necessarily be a separate commit, or is it fine since it is minor?
- Broke down `display_count`, and rewrote that part, passing the new addition `is_count_too_low` in the context. Intermediate variables `rounded_count`, `count_to_use` are newly used.

timabbott commented on web/src/filter.ts:
```diff
@@ -1697,6 +1697,10 @@ export class Filter {
     }
 
+    is_interleaved_view(): boolean {
+        return !this.can_bucket_by("channel", "topic") && !this.can_bucket_by("dm");
```
What's the purpose of this function? I'm not sure it's implementation matches the name -- any view that is a subset of a conversation view is not interleaved, for example.

timabbott replied:
I am not convinced this implementation is correct. I guess I can add a comment noting the corner cases.

timabbott replied:
`may_contain_multiple_conversations` may be a better name. The bug is that `channels:public channel:foo topic_bar` will incorrectly return false. I think you want to just check if it contains the relevant operators, without ordering.

whilstsomebody commented:
Hey @Niloth-p , are you still working on this one? If not, I would like to continue.

Niloth-p commented:
I was supposed to update this soon, but I haven't looked at this in quite a while, so I'll be glad to have you take this forward, @whilstsomebody.

whilstsomebody commented:
thanks @Niloth-p .

timabbott commented:
@Niloth-p is this ready for review? You can add the "integration review" label if so.

Niloth-p commented:
> @Niloth-p is this ready for review? You can add the "integration review" label if so.

Yes. @zulipbot add "integration review"

Niloth-p commented:
@zulipbot remove "mentor review"
@zulipbot add "integration review"

timabbott commented on web/src/unread_ops.ts:
```diff
@@ -254,33 +263,77 @@ function process_newly_read_message(
+        }
+
+        const ctx = {
+            is_count_too_low: message_count < MIN_MARK_AS_UNREAD_COUNT_LOWER_BOUND,
```
I think `show_message_count` would be a better name for this. `count` and `ctx` would also be better names than `N` and `ctx`; we don't use abbreviations in the codebase.

timabbott commented:
Posted a few comments.

Niloth-p commented:
**Updates**

> I don't understand how this check relates to the next one. You do the same thing unconditionally not far below

Oof, I had totally forgotten why I had done that. Took me a long time to realise that was not a mistake, and was a deliberate addition. I had added that early return to optimize marking entire feeds as unread, by avoiding counting the number of messages.
Not sure about the scale of the difference it would make, can't remember if I tested it or not.
I've now edited `min_message_ids_to_update` to conditionally compute itself, and removed the confusing early return.

> channels:public channel:foo topic_bar will incorrectly return false

Ah, ok, thank you! I've updated the implementation of the filter function, and added the above as a test case.
I probably didn't count that case because `is_conversation_view()` returns False for that too.

Renamed:
- is_interleaved_view -> may_contain_multiple_conversations,
- N -> count,
- is_count_too_low -> show_message_count, and switched the if-else cases to match the update.

timabbott commented on web/src/unread_ops.ts:
```diff
@@ -325,33 +334,76 @@ function process_newly_read_message(
+            const count_to_use =
+                message_count < UNREAD_COUNT_STEP_SIZE ? message_count : rounded_count;
+            display_count = `${count_to_use}+`;
+        }
```
This logic appears to be correct, but it's quite convoluted and hard to understand with so much logic in these `?` comprehensions, and uncommented complex conditionals. I'll try a refactor.

timabbott replied:
Here's what I ended up doing; this, to me, feels a lot more readable.

``` diff
diff --git a/web/src/unread_ops.ts b/web/src/unread_ops.ts
index a6a349f8d6..241793dc46 100644
--- a/web/src/unread_ops.ts
+++ b/web/src/unread_ops.ts
@@ -340,9 +340,8 @@ export function mark_as_unread_from_here(message_id: number): void {
 
     const has_found_newest = message_lists.current.data.fetch_status.has_found_newest();
     const may_contain_multiple_conversations = current_filter.may_contain_multiple_conversations();
-    let message_ids_to_update: number[] | undefined;
 
-    function do_mark_unread(): void {
+    function do_mark_unread(message_ids_to_update: number[] | undefined): void {
         // If we have already fully fetched the current view, we can
         // send the server the set of IDs to update, rather than
         // updating on the basis of the narrow.
@@ -362,49 +361,64 @@ export function mark_as_unread_from_here(message_id: number): void {
         }
     }
 
-    const min_message_ids_to_update =
-        has_found_newest || may_contain_multiple_conversations
-            ? message_lists.current
-                  .all_messages()
-                  .filter((msg) => msg.id >= message_id && !msg.unread)
-                  .map((msg) => msg.id)
-            : undefined;
-    message_ids_to_update = has_found_newest ? min_message_ids_to_update : undefined;
+    const locally_available_matching_message_ids = message_lists.current.all_messages()
+        .filter((msg) => msg.id >= message_id && !msg.unread)
+        .map((msg) => msg.id);
+    const locally_available_message_count = locally_available_matching_message_ids.length;
+    let display_count: string;
+
+    if (has_found_newest) {
+        // Since we have the anchor message ID and the newest
+        // messages, we know exactly which messages to mark as unread.
+
+        // If it's a single topic, or the number is small, we just do
+        // the request without a confirmation dialog.
+        if (!may_contain_multiple_conversations ||
+            locally_available_matching_message_ids.length < MIN_MARK_AS_UNREAD_COUNT_KNOWN) {
+            do_mark_unread(locally_available_matching_message_ids);
+            return;
+        }
 
-    if (
-        !may_contain_multiple_conversations ||
-        (message_ids_to_update !== undefined &&
-            message_ids_to_update.length < MIN_MARK_AS_UNREAD_COUNT_KNOWN)
-    ) {
-        do_mark_unread();
+        display_count = locally_available_message_count.toString();
+    } else if (locally_available_message_count < UNREAD_COUNT_STEP_SIZE ) {
+        display_count = locally_available_message_count.toString();
     } else {
-        assert(min_message_ids_to_update !== undefined);
-        const message_count = min_message_ids_to_update.length;
-        let display_count: string;
-
-        if (has_found_newest) {
-            display_count = message_count.toString();
-        } else {
-            const rounded_count =
-                Math.floor(message_count / UNREAD_COUNT_STEP_SIZE) * UNREAD_COUNT_STEP_SIZE;
-            const count_to_use =
-                message_count < UNREAD_COUNT_STEP_SIZE ? message_count : rounded_count;
-            display_count = `${count_to_use}+`;
-        }
+        // Otherwise, we round down to the nearest
+        // UNREAD_COUNT_STEP_SIZE and display as, e.g., `25+`.
+        const rounded_count =
+            Math.floor(locally_available_message_count / UNREAD_COUNT_STEP_SIZE) * UNREAD_COUNT_STEP_SIZE;
+        display_count = `${rounded_count}+`;
+    }
 
-        const ctx = {
-            show_message_count: message_count >= MIN_MARK_AS_UNREAD_COUNT_LOWER_BOUND,
-            count: display_count,
-        };
+    const context = {
+        // If we don't know how many messages will be affected, but
+        // can't prove the number is more than 10, we avoid showing a
+        // count, since it just seems weird to say "3+ messages will
+        // be marked as read".
+        //
+        // It's not obvious this case is worth having special strings
+        // for, given how unlikely it is. A sample scenario is that
+        // we're be that we're near the fetched bottom of a /near/1
+        // search view where can_apply_locally is false, which will
+        // have triggered a request for the next batch of messages
+        // from the server, but that request has not returned. But it
+        // may happen more offline if the client is intermittantly
+        // offline.
+        show_message_count: locally_available_message_count >= MIN_MARK_AS_UNREAD_COUNT_LOWER_BOUND,
+        count: display_count,
+    };
 
-        confirm_dialog.launch({
-            html_heading: $t_html({defaultMessage: "Mark messages as unread?"}),
-            html_body: render_confirm_mark_as_unread_from_here(ctx),
-            on_click() {
-                do_mark_unread();
-            },
-        });
-    }
+    confirm_dialog.launch({
+        html_heading: $t_html({defaultMessage: "Mark messages as unread?"}),
+        html_body: render_confirm_mark_as_unread_from_here(context),
+        on_click() {
+            if (has_found_newest) {
+                do_mark_unread(locally_available_matching_message_ids);
+            } else {
+                do_mark_unread(undefined);
+            }
+        },
+    });
 }
 
 function do_mark_unread_by_narrow(
```

timabbott commented:
I pushed an updated version that I believe has equivalent logic (and seemed to in manual testing, though please review and do a round yourself just in case). This, to me, feels a lot more readable.
